src/session/liveliness.c

```diff
@@ -0,0 +1,340 @@
+    _z_liveliness_pending_query_t *dst = z_malloc(sizeof(_z_liveliness_pending_query_t));
+    if (dst != NULL) {
+        _z_liveliness_pending_query_copy(dst, src);
+    }
+    return dst;
+}
+
+uint32_t _z_liveliness_get_query_id(_z_session_t *zn) { return zn->_liveliness_query_id++; }
+
+z_result_t _z_liveliness_register_pending_query(_z_session_t *zn, uint32_t id, _z_liveliness_pending_query_t *pen_qry) {
+    z_result_t ret = _Z_RES_OK;
+
+    _Z_DEBUG("Register liveliness query for (%ju:%.*s)", (uintmax_t)pen_qry->_key._id,
+             (int)_z_string_len(&pen_qry->_key._suffix), _z_string_data(&pen_qry->_key._suffix));
+
+    _zp_session_lock_mutex(zn);
+
+    const _z_liveliness_pending_query_t *pq =
+        _z_liveliness_pending_query_intmap_get(&zn->_liveliness_pending_queries, id);
+    if (pq != NULL) {
+        _Z_ERROR("Duplicate liveliness query id %i", (int)id);
+        ret = _Z_ERR_ENTITY_DECLARATION_FAILED;
+    } else {
+        _z_liveliness_pending_query_intmap_insert(&zn->_liveliness_pending_queries, id,
+                                                  _z_liveliness_pending_query_clone(pen_qry));
+    }
+
+    _zp_session_unlock_mutex(zn);
+
+    return ret;
+}
+
+z_result_t _z_liveliness_pending_query_reply(_z_session_t *zn, uint32_t interest_id, const _z_keyexpr_t keyexpr,
+                                             const _z_timestamp_t *timestamp) {
+    z_result_t ret = _Z_RES_OK;
+
+    _zp_session_lock_mutex(zn);
+
+    const _z_liveliness_pending_query_t *pq =
+        _z_liveliness_pending_query_intmap_get(&zn->_liveliness_pending_queries, interest_id);
+    if (pq == NULL) {
+        ret = _Z_ERR_ENTITY_UNKNOWN;
+    }
+
+    _Z_DEBUG("Liveliness pending query reply %i resolve result %i", (int)interest_id, ret);
+
+    if (ret == _Z_RES_OK) {
+        _Z_DEBUG("Resolving %d - %.*s on mapping 0x%x", keyexpr._id, (int)_z_string_len(&keyexpr._suffix),
+                 _z_string_data(&keyexpr._suffix), _z_keyexpr_mapping_id(&keyexpr));
+        _z_keyexpr_t expanded_ke = __unsafe_z_get_expanded_key_from_key(zn, &keyexpr);
+        _Z_DEBUG("Reply liveliness query for %d - %.*s", expanded_ke._id, (int)_z_string_len(&expanded_ke._suffix),
+                 _z_string_data(&expanded_ke._suffix));
+
+        if (!_z_keyexpr_suffix_intersects(&pq->_key, &expanded_ke)) {
+            ret = _Z_ERR_QUERY_NOT_MATCH;
+        }
+
+        if (ret == _Z_RES_OK) {
+            _z_encoding_t encoding = _z_encoding_null();
+            _z_reply_t reply = _z_reply_create(expanded_ke, zn->_local_zid, _z_bytes_null(), timestamp, &encoding,
+                                               Z_SAMPLE_KIND_PUT, _z_bytes_null());
+
+            pq->_callback(&reply, pq->_arg);
+            _z_reply_clear(&reply);
+        }
+    }
+
+    if (ret == _Z_RES_OK) {
+        _z_liveliness_pending_query_intmap_remove(&zn->_liveliness_pending_queries, interest_id);
```

Comment:
If we remove the pending query from the list here, we won't be able to process next DeclareTokens, right ?